storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java

```diff
@@ -157,13 +157,20 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate metada
         RemoteLogSegmentState targetState = metadataUpdate.state();
         RemoteLogSegmentId remoteLogSegmentId = metadataUpdate.remoteLogSegmentId();
         RemoteLogSegmentMetadata existingMetadata = idToSegmentMetadata.get(remoteLogSegmentId);
+        if (!isInitialized() && existingMetadata == null) {
+            log.debug("Dropping the event: {} as the base-metadata about the segment was already removed", metadataUpdate);
```

Comment:
Do you think this should be a WARN or at least INFO log, since this is an issue if some unexpected event is coming?

Reply:
If the logger level is changed to INFO, then it will print a log statement for each event on startup/restart which can quickly pollute the logs. The idea taken here is not to print a log when the partition is uninitialised. Once the partition gets initialised, any unexpected event transition will be logged as ERROR.

Reply:
Ah, I see. if block will be true when consumer is still initializing to consume and process metadata records from the internal topic, and some records lost their earlier records because of log retention. Is that right? 

I think the current solution makes sense. But about the assumption:
> By the time, the segments of __remote_log_metadata gets expired, the actual user-topic segments should already be deleted.

I think we need to file a JIRA to improve this assumption. We can't just rely on it and say it's user's fault. WDYT?

Reply:
I am unclear as to why you have this assumption? Won't the code work even if a customer has infinite retention for the __remote_log_metadata? Won't it work even if the retention of __remote_log_metadata is smaller than the data in the topics? Worst case scenario metadata about the segments gets deleted which means that the segments are unreachable (both for reads and deletes) but data is not lost, no? I am with Luke on filing a JIRA for a "nicer" fix, but I think historically Kafka never modified internal topics once created.

Reply:
Filed KAFKA-15682 to followup on this.